examples/sudoku_zkp/src/main.rs

```diff
@@ -0,0 +1,254 @@
+use sunscreen::{
+    types::zkp::{ConstrainCmp, NativeField},
+    zkp_program, BackendField, Compiler, Runtime, ZkpProgramInput,
+};
+use sunscreen_zkp_backend::{bulletproofs::BulletproofsBackend, ZkpBackend};
+
+type BPField = NativeField<<BulletproofsBackend as ZkpBackend>::Field>;
+
+fn main() {
+    let app = Compiler::new()
+        .zkp_backend::<BulletproofsBackend>()
+        .zkp_program(sudoku_proof)
+        .compile()
+        .unwrap();
+
+    let prog = app.get_zkp_program(sudoku_proof).unwrap();
+
+    let runtime = Runtime::new_zkp(&BulletproofsBackend::new()).unwrap();
+
+    let ex_puzzle = [
+        [0, 7, 0, 0, 2, 0, 0, 4, 6],
+        [0, 6, 0, 0, 0, 0, 8, 9, 0],
+        [2, 0, 0, 8, 0, 0, 7, 1, 5],
+        [0, 8, 4, 0, 9, 7, 0, 0, 0],
+        [7, 1, 0, 0, 0, 0, 0, 5, 9],
+        [0, 0, 0, 1, 3, 0, 4, 8, 0],
+        [6, 9, 7, 0, 0, 2, 0, 0, 8],
+        [0, 5, 8, 0, 0, 0, 0, 6, 0],
+        [4, 3, 0, 0, 8, 0, 0, 7, 0],
+    ];
+
+    let ex_sol = [
+        [8, 7, 5, 9, 2, 1, 3, 4, 6],
+        [3, 6, 1, 7, 5, 4, 8, 9, 2],
+        [2, 4, 9, 8, 6, 3, 7, 1, 5],
+        [5, 8, 4, 6, 9, 7, 1, 2, 3],
+        [7, 1, 3, 2, 4, 8, 6, 5, 9],
+        [9, 2, 6, 1, 3, 5, 4, 8, 7],
+        [6, 9, 7, 4, 1, 2, 5, 3, 8],
+        [1, 5, 8, 3, 7, 9, 2, 6, 4],
+        [4, 3, 2, 5, 8, 6, 9, 7, 1],
+    ];
+
+    let board: Vec<ZkpProgramInput> = vec![ex_sol.map(|a| a.map(BPField::from)).into()];
+
+    let cons: Vec<ZkpProgramInput> = vec![ex_puzzle.map(|a| a.map(BPField::from)).into()];
+
+    let proof = runtime.prove(prog, cons.clone(), vec![], board).unwrap();
+
+    let verify = runtime.verify(prog, &proof, cons.clone(), vec![]);
+
+    assert!(verify.is_ok());
+}
+
+#[zkp_program(backend = "bulletproofs")]
+fn sudoku_proof<F: BackendField>(
+    #[constant] constraints: [[NativeField<F>; 9]; 9],
+    board: [[NativeField<F>; 9]; 9],
+) {
+    fn contains_circuit<F: BackendField>(
+        arr: [ProgramNode<NativeField<F>>; 9],
+        i: NativeField<F>,
+    ) -> ProgramNode<NativeField<F>> {
+        let mut circuit = NativeField::<F>::from(1).into_program_node();
+        for a in arr {
+            circuit = circuit * (i.into_program_node() - a);
```

Comment:
I would probably do the `for i in 1..=9` iteration in this function. Rename this helper `assert_unique_numbers` and add a comment that this function checks that the values 1 through 9 each appear once in `arr`.

Reply:
And add the `.constrain_eq(0)` in here.